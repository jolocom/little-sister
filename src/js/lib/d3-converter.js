import rdf from 'rdflib'
let FOAF = rdf.Namespace('http://xmlns.com/foaf/0.1/')
let DC = rdf.Namespace('http://purl.org/dc/terms/')
let RDF = rdf.Namespace('http://www.w3.org/1999/02/22-rdf-syntax-ns#')
import STYLES from 'styles/app.js'

// D3 Converter takes a node (a node in this context is an array of triples that
// describe an rdf document) and then, based on that returns an array where the
// triples are represented in a different format. This array can then be fed
// to D3 to draw a graph based on the data

class D3Converter {
  convertToD3(node, i, n) {
    // We need to know the index of the node and the total amount of nodes
    // in order to be able to calculate their initial position, so that they are
    // possitioned in a circle
    this.i = i + 1
    this.n = n

    let uri = node.uri

    let props = {
      uri: null,
      name:null,
      description:null,
      img:null,
      type:null,
      rank: null,
      x: null,
      y: null
    }

    // We create a rdf.graph() object, and populate it with the triples, this
    // allows us to then parse them using the rdflib's function
    // rdf.graph().statementsMatching()

    let g = rdf.graph()
    for (let i = 0; i < node.length; i++) {
      g.add(node[i].subject, node[i].predicate, node[i].object)
    }
    // Calculating the coordinates of the nodes so we can put them in a circle
    let angle = (2 * Math.PI) / this.n
    let halfwidth = STYLES.width / 2
    let halfheight = STYLES.height / 2

    props.x = Math.sin(angle * this.i) * halfwidth + halfwidth
    props.y =(Math.cos(angle * this.i) * halfwidth + halfheight)

    // Updating the attributes of the node object. The resulting object will have
    // all of it's props filled in, and will be ready to be rendered by D3
    // Note, if a triple is not present, it will be set to null.
    props.uri = uri

    // If the resource is a URI, it's value is stored next to the 'uri' key in the object
    // otherwise it's value is stored in the 'value' key of the object. We need to make
    // sure we are assigning the value regardless of where it's stored
    let name = g.statementsMatching(undefined, FOAF('name'), undefined)
    if (name.length > 0) props.name = name[0].object.value ? name[0].object.value : name[0].object.uri
    else props.name = null

    let description = g.statementsMatching(undefined, DC('description'), undefined)
    if (description.length > 0) props.description = description[0].object.value ? description[0].object.value : description[0].object.uri
    else props.description = null

    let type = g.statementsMatching(undefined, RDF('type'), undefined)
    if (type.length > 0) props.type = type[0].object.value ? type[0].object.value : type[0].object.uri
    else props.type = null

  //This labels the nodes as center and adjacent, a bit hacky, but it works allright
    if (i) props.rank = 'adjacent'
    if (!i) props.rank = 'center'

    return props
  }
}

export default D3Converter
