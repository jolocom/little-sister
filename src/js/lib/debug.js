import debug from 'debug'
import {flow, map} from 'lodash/fp'

debug.log = console.trace.bind(console)

// Prints an array of triples using console.table
console.rdftable = (array) => {
  flow(
      map((statement) => ({
        subject: statement.subject.uri || statement.subject.value,
        predicate: statement.predicate.uri.split("/").slice(-1)[0],
        object: statement.object.uri || statement.object.value
      })),
      console.table.bind(console))
  (array)
}

export default debug