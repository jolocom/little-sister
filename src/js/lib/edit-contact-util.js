import Immutable from 'immutable'

export const validateChanges = (state) => {
  const {newInformation, originalInformation: original} = state
    .get('information').toJS()
  const newFields = [...newInformation.emails, ...newInformation.phones]
  const oldFields = [...original.emails, ...original.phones]

  const isValid = newFields.every(e => (e.blank || e.valid || e.delete)) &&
    oldFields.every(e => e.delete || !e.update || (e.valid && !e.verified))

  return state.setIn(['showErrors'], !isValid)
}

const isBlank = (value, field) => {
  if (typeof value === 'object') {
    return !value.value || !value.value.trim()
  }
  return !value.trim()
}

const isValidValue = (value, field) => {
  switch (field) {
    case 'phones':
      return (/^([\d.]+)$/.test(value.value) ||
        /^\+([\d]+)$/.test(value.value))
    case 'emails':
      return /^([\w.]+)@([\w.]+)\.(\w+)/.test(value)
    default:
      return false
  }
}

const parseActionValueToObject = (value, field) => field === 'phones'
  ? value : {value}

const addTypeAttribute = (field) => field === 'phones'
  ? {type: 'personal'} : {}

export const setNewFieldValue = (state, {field, index, value}) => state.mergeIn(
  ['information', 'newInformation', field, index], {
    ...parseActionValueToObject(value, field),
    valid: isValidValue(value, field),
    blank: isBlank(value, field)
  })

export const mapAccountInformationToState = ({email, phone, addresses = []}) =>
  Immutable.fromJS({
    loading: false,
    showErrors: false,
    information: {
      newInformation: {
        emails: [email.length > 0 ? {delete: true} : {
          value: '',
          delete: false,
          blank: true,
          valid: true
        }],
        phones: [phone.length > 0 ? {delete: true} : {
          value: '',
          type: 'personal',
          blank: true,
          delete: false,
          valid: true
        }],
        addresses: [addresses.length > 0 ? {delete: true} : {
          streetWithNumber: {value: '', valid: true},
          zip: {value: '', valid: true},
          city: {value: '', valid: true},
          state: {value: '', valid: true},
          country: {value: '', valid: true},
          delete: false,
          blank: true,
          valid: true
        }]
      },
      originalInformation: {
        emails: email.map(({address, verified, id}) => ({
          value: address,
          verified,
          id,
          delete: false,
          update: false,
          valid: true
        })),
        phones: phone.map(({number, verified, id}) => ({
          value: number,
          // TODO: phone type needs to be delivered from the backend
          type: 'personal',
          verified,
          id,
          delete: false,
          update: false,
          valid: true
        })),
        addresses: addresses.map(({streetWithNumber, zip, city, state, country}) => ({ // eslint-disable-line max-len
          streetWithNumber: {...streetWithNumber, valid: true},
          zip: {...zip, valid: true},
          city: {...city, valid: true},
          state: {...state, valid: true},
          country: {...country, valid: true},
          delete: false,
          blank: true,
          valid: true
        }))
      }
    }
  })

const canAddNewField = (state, field, index) => {
  if (index === 0) { return true }
  if (!state
    .getIn(['information', 'newInformation', field, index - 1, 'delete'])) {
    return !state
      .getIn(['information', 'newInformation', field, index - 1, 'blank'])
  }
  return canAddNewField(state, field, index - 1)
}

export const addNewField = (state, {field, index}) => {
  if (field === 'addresses') {
    return state.mergeIn(['information', 'newInformation', field, index], {
      streetWithNumber: {value: '', valid: true},
      zip: {value: '', valid: true},
      city: {value: '', valid: true},
      state: {value: '', valid: true},
      country: {value: '', valid: true},
      delete: false,
      blank: true,
      valid: true
    })
  }
  if (canAddNewField(state, field, index)) {
    return state.mergeIn(['information', 'newInformation', field, index], {
      ...addTypeAttribute(field), value: '', verified: false, valid: false,
      delete: false, blank: true
    })
  }
  return state
}

const canUpdate = (state, field, index) => !state.getIn(
  ['information', 'originalInformation', field, index, 'verified']) &&
  state.getIn(['information', 'originalInformation', field]).toJS().length > 0

export const updateOriginalValue = (state, {field, value, index}) => {
  if (canUpdate(state, field, index)) {
    return state.mergeIn(['information', 'originalInformation', field, index], {
      ...parseActionValueToObject(value, field),
      valid: isValidValue(value, field),
      delete: value === '',
      update: true
    })
  }
  return state
}
const collectChages =
  (state, {remove, update, set}, key, webId, attributeType) => [].concat(
    state.originalInformation[key].map(
      e => e.delete ? remove(webId, attributeType, e.id)
      : (e.update && e.valid && !e.verified)
      ? update(webId, attributeType, e.id, e.value)
      : null
    ), state.newInformation[key].map(
      e => (e.delete || e.blank || !e.valid)
      ? null : set(webId, e.value)
  ))

export const submitChanges = (backend, services, state, webId) => {
  let solidAgent = backend.solid
  const emailOperations = {
    set: solidAgent.setEmail.bind(solidAgent),
    remove: solidAgent.deleteEntry.bind(solidAgent),
    update: solidAgent.updateEntry.bind(solidAgent)
  }
  const phoneOperations = {
    set: solidAgent.setPhone.bind(solidAgent),
    remove: solidAgent.deleteEntry.bind(solidAgent),
    update: solidAgent.updateEntry.bind(solidAgent)
  }
  let promises = [].concat(
    collectChages(state, emailOperations, 'emails', webId, 'email'),
    collectChages(state, phoneOperations, 'phones', webId, 'phone'))
  return Promise.all(promises)
}
