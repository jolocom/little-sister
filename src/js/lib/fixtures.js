export default {}

export let bankUri = 'https://localhost:8443'

export let contacts = [{
  username: 'eelco',
  name: 'Eelco Wiersma',
  email: 'hi@eelcowiersma.nl'
}, {
  username: 'joachim',
  name: 'Joachim Lohkamp',
  email: 'joachim@jolocom.com'
}, {
  username: 'justas',
  name: 'Justas Azna',
  email: 'justas@jolocom.com'
}, {
  username: 'christian',
  name: 'Christian Hildebrand',
  email: 'christian@jolocom.com'
}, {
  username: 'anna',
  name: 'Anna Blume',
  email: 'anna@jolocom.com'
}]

export let conversations = [{
  type: 'user',
  username: 'justas',
  items: [{
    author: {
      username: 'justas',
      name: 'Justas Azna'
    },
    content: 'This is completely fucked up!',
    date: new Date()
  }]
}, {
  type: 'user',
  username: 'joachim',
  items: [{
    author: {
      username: 'joachim',
      name: 'Joachim Lohkamp'
    },
    content: 'How are you?',
    date: new Date()
  }]
}]
