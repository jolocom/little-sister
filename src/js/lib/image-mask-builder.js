export default class ImageMaskBuilder {
  constructor() {
    this.paths = []
    this.currentPath = null
  }

  startNewPath() {
    this.currentPath = []
    this.paths.push(this.currentPath)
  }

  endPath() {
    this.currentPath = null
  }

  addPoint(x, y) {
    // if (x === 0 && y === 0) {
    //   return
    // }

    this.currentPath.push([x, y])
  }
}
