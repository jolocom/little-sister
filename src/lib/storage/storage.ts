import {
  createConnection,
  ConnectionOptions,
  Connection,
} from 'typeorm/browser'
import { plainToClass } from 'class-transformer'
import {
  PersonaEntity,
  MasterKeyEntity,
  VerifiableCredentialEntity,
  SignatureEntity,
  CredentialEntity,
} from 'src/lib/storage/entities'
import { SignedCredential } from 'jolocom-lib/js/credentials/signedCredential/signedCredential'

interface PersonaAttributes {
  did: string
  controllingKeyPath: string
}

interface EncryptedSeedAttributes {
  encryptedEntropy: string
  timestamp: number
}

interface ModifiedCredentialEntity {
  propertyName: string
  propertyValue: string[]
  verifiableCredential: VerifiableCredentialEntity
}

export class Storage {
  private connection!: Connection
  private config: ConnectionOptions

  store = {
    verifiableCredential: this.storeVClaim.bind(this),
    persona: this.storePersonaFromJSON.bind(this),
    encryptedSeed: this.storeEncryptedSeed.bind(this),
  }

  get = {
    persona: this.getPersonas.bind(this),
    verifiableCredential: this.getVCredential.bind(this),
    attributesByType: this.getAttributesByType.bind(this),
    vCredentialsByAttributeValue: this.getVCredentialsForAttribute.bind(this),
    encryptedSeed: this.getEncryptedSeed.bind(this),
  }

  delete = {
    verifiableCredential: this.deleteVCred.bind(this),
  }

  constructor(config: ConnectionOptions) {
    this.config = config
  }

  private async createConnectionIfNeeded(): Promise<void> {
    if (!this.connection) {
      this.connection = await createConnection(this.config)
    }
  }

  // TODO: refactor needed on multiple personas
  private async getPersonas(query?: object): Promise<PersonaEntity[]> {
    await this.createConnectionIfNeeded()
    return this.connection.manager.find(PersonaEntity)
  }

  private async getVCredential(query?: object): Promise<SignedCredential[]> {
    await this.createConnectionIfNeeded()
    const entities = await this.connection.manager.find(
      VerifiableCredentialEntity,
      {
        where: query,
        relations: ['claim', 'proof', 'subject'],
      },
    )

    return entities.map(e => e.toVerifiableCredential())
  }

  private async getAttributesByType(type: string[]) {
    await this.createConnectionIfNeeded()
    const localAttributes = await this.connection
      .getRepository(CredentialEntity)
      .createQueryBuilder('credential')
      .leftJoinAndSelect(
        'credential.verifiableCredential',
        'verifiableCredential',
      )
      .where('verifiableCredential.type = :type', { type })
      .getMany()

    const results = this.groupAttributesByCredentialId(localAttributes).map(
      entry => ({
        verification: entry.verifiableCredential.id,
        values: entry.propertyValue,
        fieldName: entry.propertyName,
      }),
    )

    return { type, results }
  }

  // TODO rework
  private groupAttributesByCredentialId(
    attributes: CredentialEntity[],
  ): ModifiedCredentialEntity[] {
    // Convert values to arrays for easier concatination later
    const modifiedAttributes = attributes.map(attr => ({
      ...attr,
      propertyValue: [attr.propertyValue],
    }))

    // Helper function
    const findByCredId = (
      arrToSearch: ModifiedCredentialEntity[],
      value: ModifiedCredentialEntity,
    ) =>
      arrToSearch.findIndex(
        entry =>
          entry.verifiableCredential.id === value.verifiableCredential.id,
      )

    return modifiedAttributes.reduce(
      (acc: ModifiedCredentialEntity[], curr: ModifiedCredentialEntity) => {
        const matchingIndex = findByCredId(acc, curr)

        if (matchingIndex >= 0) {
          acc[matchingIndex].propertyValue = [
            ...acc[matchingIndex].propertyValue,
            ...curr.propertyValue,
          ]
          return acc
        } else {
          return [...acc, curr]
        }
      },
      [],
    )
  }

  private async getVCredentialsForAttribute(
    attribute: string,
  ): Promise<SignedCredential[]> {
    await this.createConnectionIfNeeded()
    const entities = await this.connection
      .getRepository(VerifiableCredentialEntity)
      .createQueryBuilder('verifiableCredential')
      .leftJoinAndSelect('verifiableCredential.claim', 'claim')
      .leftJoinAndSelect('verifiableCredential.proof', 'proof')
      .leftJoinAndSelect('verifiableCredential.subject', 'subject')
      .where('claim.propertyValue = :attribute', { attribute })
      .getMany()

    return entities.map(e => e.toVerifiableCredential())
  }

  private async getEncryptedSeed(): Promise<string> {
    await this.createConnectionIfNeeded()
    const masterKeyEntity = await this.connection.manager.find(MasterKeyEntity)
    return masterKeyEntity[0].encryptedEntropy
  }

  private async storePersonaFromJSON(args: PersonaAttributes): Promise<void> {
    await this.createConnectionIfNeeded()
    const persona = plainToClass(PersonaEntity, args)
    await this.connection.manager.save(persona)
  }

  private async storeEncryptedSeed(
    args: EncryptedSeedAttributes,
  ): Promise<void> {
    await this.createConnectionIfNeeded()
    const encryptedSeed = plainToClass(MasterKeyEntity, args)
    await this.connection.manager.save(encryptedSeed)
  }

  private async storeVClaim(vCred: SignedCredential): Promise<void> {
    await this.createConnectionIfNeeded()
    const verifiableCredential = VerifiableCredentialEntity.fromVerifiableCredential(
      vCred,
    )

    const signature = SignatureEntity.fromLinkedDataSignature(vCred.proof)

    const claims = CredentialEntity.fromVerifiableCredential(vCred)
    claims.forEach(claim => (claim.verifiableCredential = verifiableCredential))

    signature.verifiableCredential = verifiableCredential

    verifiableCredential.proof = [signature]
    verifiableCredential.claim = claims

    await this.connection.manager.save(verifiableCredential)
  }

  private async deleteVCred(id: string): Promise<void> {
    await this.createConnectionIfNeeded()
    await this.connection.manager
      .createQueryBuilder()
      .delete()
      .from(CredentialEntity)
      .where('verifiableCredential = :id', { id })
      .execute()

    await this.connection.manager
      .createQueryBuilder()
      .delete()
      .from(SignatureEntity)
      .where('verifiableCredential = :id', { id })
      .delete()
      .execute()

    await this.connection.manager
      .createQueryBuilder()
      .delete()
      .from(VerifiableCredentialEntity)
      .where('id = :id', { id })
      .execute()
  }
}
