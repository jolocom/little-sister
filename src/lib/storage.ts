const SQLite = require('react-native-sqlite-storage')
import { dbHelper, TableOptions } from 'src/lib/dbHelper'
import { Location, ResultSet, Transaction } from 'react-native-sqlite-storage'

interface AsyncTransaction {
  executeSql: (query: string, args?: any[]) => Promise<{result: ResultSet, transaction: Transaction}>
}

interface SQLiteDatabase {
  enablePromise: (flag: boolean) => void
  DEBUG: (flag: boolean) => void
  openDatabase: (args: {name: string, location: Location}) => Promise<SQLiteDatabase>
  executeSql: (query: string) => Promise<void>
  executeQuery: (db: SQLiteDatabase, query: string) => Promise<void>
  close: (successCB: Function, errCB: Function) => Promise<void>
  transaction: (tx: Function) => Promise<ResultSet>
}

export class Storage {
  private sqlLite: SQLiteDatabase = SQLite
  private dbName = 'LocalSmartWalletData'
  private location: Location = 'default'

  constructor() {
    this.enablePromise()
    // this.sqlLite.DEBUG(false)
  }

  private enablePromise() : void {
    this.sqlLite.enablePromise(true)
  }

  private async getDbInstance() : Promise<SQLiteDatabase> {
    const db = await this.sqlLite.openDatabase({
      name: this.dbName,
      location: this.location
    })

    await db.executeSql('PRAGMA foreign_keys = ON;')
    return db
  }

  async closeDB(db: SQLiteDatabase) : Promise<{}> {
    return new Promise((resolve, reject) => {
      return db.close(resolve, reject)
    })
  }

  async provisionTables() : Promise<void> {
    const tableData: TableOptions[] = dbHelper.defaultTables
    const db = await this.getDbInstance()
    await Promise.all(tableData.map(async t => await this.createTable(t, db)))
    await this.closeDB(db)
  }

  async addPersona(args: {did: string, controllingKey: number}) {
    const db = await this.getDbInstance()
    const query = dbHelper.addPersonaQuery(args)
    await this.executeQuery(db, query)
    await this.closeDB(db)
  }

  async addMasterKey(entropy: string) {
    const db = await this.getDbInstance()
    const query = dbHelper.addMasterKeyQuery(entropy)
    await this.executeQuery(db, query)
    await this.closeDB(db)
  }

  async addDerivedKey(args: { encryptedWif: string, path: string, entropySource: number, keyType: string }) {
    const db = await this.getDbInstance()
    const query = dbHelper.addDerivedKeyQuery(args)
    await this.executeQuery(db, query)
    await this.closeDB(db)
  }

  private async createTable(options : TableOptions, db: SQLiteDatabase) : Promise<void> {
    const query = dbHelper.createTableQuery(options)
    await this.executeQuery(db, query)
  }

  private async executeQuery(db: SQLiteDatabase, query: string) : Promise<ResultSet> {
    return await db.transaction(async (tx: AsyncTransaction) => {
      const response = await tx.executeSql(query, [])
      return response.result
    })
  }
}
