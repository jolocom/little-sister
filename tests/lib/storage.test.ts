import { Storage }  from 'src/lib/storage'

describe('Storage lib', () => {
  let StorageAgent

  beforeEach(() => {
    StorageAgent = Storage
  })

  it('should initialize correctly', () => {
    const mock = StorageAgent.prototype.enablePromise = jest.fn()

    new StorageAgent()
    expect(mock).toHaveBeenCalledTimes(1)
  })

  it('should attempt to provision tables', async () => {
    const agent = new StorageAgent()

    const mockExecQuery = agent.executeQuery = jest.fn()
    const mockCloseDb = agent.closeDB = jest.fn().mockResolvedValue()

    const mockDbInstance = { transaction: jest.fn() }
    const mockGetDbInstnace = agent.getDbInstance = jest.fn()
      .mockResolvedValue(mockDbInstance)

    await agent.provisionTables()

    expect(mockExecQuery.mock.calls).toMatchSnapshot()
    expect(mockGetDbInstnace).toHaveBeenCalledTimes(1)
    expect(mockCloseDb).toHaveBeenCalledTimes(1)
  })
})
