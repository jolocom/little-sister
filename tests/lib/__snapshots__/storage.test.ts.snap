// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storage Lib | dbHelper should correctly assemble table creation query 1`] = `"CREATE TABLE IF NOT EXISTS Personas (did VARCHAR(20) NOT NULL PRIMARY KEY, FOREIGN KEY(controllingKey) REFERENCES Keys(id) )"`;

exports[`Storage Lib | dbHelper should correctly assemble table creation query 2`] = `"CREATE TABLE IF NOT EXISTS Keys (id INTEGER PRIMARY KEY NOT NULL UNIQUE, FOREIGN KEY(entropySource) REFERENCES MasterKeys(id) )"`;

exports[`Storage lib should attempt to provision tables 1`] = `
Array [
  Array [
    Object {
      "transaction": [MockFunction],
    },
    "CREATE TABLE IF NOT EXISTS Personas (did VARCHAR(20) NOT NULL UNIQUE COLLATE NOCASE PRIMARY KEY, controllingKey INTEGER , FOREIGN KEY(controllingKey) REFERENCES Keys(id) )",
  ],
  Array [
    Object {
      "transaction": [MockFunction],
    },
    "CREATE TABLE IF NOT EXISTS Keys (id INTEGER PRIMARY KEY NOT NULL UNIQUE, wif VARCHAR(20) UNIQUE NOT NULL, path VARCHAR(10) NOT NULL, entropySource INTEGER , algorithm TEXT NOT NULL, FOREIGN KEY(entropySource) REFERENCES MasterKeys(id) )",
  ],
]
`;
